### DISCUSSION TITLE
Parallelization of ca.jacobian for large Function.map / collocated problems

### DISCUSSION THREAD

**Discussion by jackvreeken** [casadi/casadi#3323] [Q&A] [answered]
When I have a `f.map(N)` problem, CasADi will (apparently) paralllelize calculating its jacobian when calling e.g. nlpsol. But when calculating the Jacobian using `ca.jacobian` it will not. Searching a little this seems to also be an unanswered (follow-up?) question in https://github.com/casadi/casadi/discussions/3197 .

It's a rather big set of equations I'm testing with right now, with ~1 million constraints and ~400 k variables. It takes about 2.5 minutes to calculate the linear coefficient matrix. The quickest way I can do it now is by `expand()`ing first, and then use Python multiprocessing to calculate `ca.jacobian(slice_of_g, x)` (g sliced into 16 pieces for my PC). That cuts it down to about 16 seconds. So I would imagine that CasADi itself should be able to be faster still if it itself would parallelize.

**Comment by charleskhazoom** [comment]
To enable parallelization using `map`, f.map(N,'thread',nb_of_threads) worked for me in Matlab.

If you are using codegen, you can try 'openmp' option instead of 'thread' and compile with the  `-fopenmp` flag

**Comment by jaeandersson** [comment, marked as answer]
My guess is that it has something to do with ```casadi::jacobian``` performing the Jacobian calculation on an MX graph instead of using the ```FunctionInternal::get_jacobian``` for the map object. I wouldn't be able to know for sure or be able to suggest a fix without diving into the code. In general, the heuristics for Jacobian calculation are complex with plenty of room for improvement all over.

**Reply by jackvreeken** [reply]
Alright, I'm already happy that it might actually make sense to see this behavior :) I'll keep my workaround of `multiprocessing` for now then, but good to know that there is probably a lot of room for improvement.

Can you also take a guess at how (if at all) `qpsol` is affected? It does `M::jacobian(g, x)`, and I'm not sure if that's the same thing as me calling `ca.jacobian` in Python.

**Reply by jaeandersson** [reply]
Yes, ```M::jacobian(g, x)``` should be the same as `casadi.jacobian` in Python. So probably it affects `qpsol` the same way. On the other hand, the evaluation only needs to be done once with `qpsol`, to get the linearization. Not in each iteration of the solver. To work around it, you could probably use `casadi.conic` instead of `casadi.qpsol`.

**Reply by jackvreeken** [reply]
OK, understood! I'll see if I can test with `conic`, but since I need the matrix `A` myself[0] I'll probably keep using `multiprocessing`.

[0] Dealing with a multi-objective problem, and don't want to recalculate the jacobian of 1M constraints for every objective (just that of one additional one).

**Comment by jgillis** [comment]
Can you provide an example?

**Reply by jackvreeken** [reply]
Dove in a little deeper than I wanted to, but seems as if some parts of what we're doing in RTC-Tools makes calculating the Jacobian quite slow. Anyway, slow or fast, it's still serial, which is what this discussion was about.

Test case (LZMA compressed or it doesn't fit): [jacobian_speed_test_case.zip](https://github.com/casadi/casadi/files/12371587/jacobian_speed_test_case.zip)


Two notebooks in there, one with the full problem ("jacobian_speed.ipynb"), and one where I look at just the map and when it gets slow. As said above, in all cases, serial evaluation even though there is an OpenMP (or thread) `f.map`.
